tests/test_kubernetes.py

```diff
@@ -0,0 +1,166 @@
+from collections import namedtuple
+import pytest
+import fv3config
+
+MemoryArgs = namedtuple(
+    'MemoryArgs', ['memory_gb', 'memory_gb_limit_in', 'memory_gb_limit']
+)
+
+
+@pytest.fixture
+def config_location():
+    return 'gs://my-bucket/fv3config.yml'
+
+
+@pytest.fixture
+def outdir():
+    return 'gs://my-bucket/rundir'
+
+
+@pytest.fixture(params=['default', 'tagged'])
+def docker_image(request):
+    if request.param == 'default':
+        return 'us.gcr.io/vcm-ml/fv3gfs-python'
+    elif request.param == 'tagged':
+        return 'us.gcr.io/vcm-ml/fv3gfs-python:1.0'
+
+
+@pytest.fixture(params=['gcloud', 'local'])
+def runfile(request):
+    if request.param == 'gcloud':
+        return 'gs://my-bucket/runfile.py'
+    if request.param == 'local':
+        return '/inside/docker/dir/runfile.py'
+
+
+@pytest.fixture
+def jobname():
+    return 'my-job-name'
+
+
+@pytest.fixture(params=['my-project', None])
+def google_cloud_project(request):
+    return request.param
+
+
+@pytest.fixture(params=["default", "manual_limit"])
+def memory_args(request):
+    return {
+        "default": MemoryArgs(3.6, None, 3.6),
+        "manual_limit": MemoryArgs(3.6, 4.0, 4.0),
+    }[request.param]
+
+
+@pytest.fixture(params=["1_cpu", "2.5_cpu"])
+def cpu_count(request):
+    if request.param == "1_cpu":
+        return 1
+    elif request.param == "2.5_cpu":
+        return 2.5
+
+
+@pytest.fixture(params=['my-secret-name', None])
+def gcp_secret(request):
+    return request.param
+
+
+@pytest.fixture(
+    params=['Always', 'IfNotPresent']
+)
+def image_pull_policy(request):
+    return request.param
+
+
+def test_local_config_rejected(outdir, docker_image, runfile):
+    with pytest.raises(ValueError):
+        fv3config.fv3run._kubernetes._get_job(
+            '/local/dir/config.yml', outdir, docker_image, runfile)
+
+
+def test_local_outdir_rejected(config_location, docker_image, runfile):
+    with pytest.raises(ValueError):
+        fv3config.fv3run._kubernetes._get_job(
+            config_location, '/local/outdir', docker_image, runfile)
+
+
+def test_get_job(
+        config_location, outdir, docker_image, runfile, jobname,
```

Comment:
This test is bit tricky to read, some whitespace and comments would help, but it would be better to split it into several functions. There are a lot of assert statements for one tests.

Reply:
Hahaha you should have seen it before I refactored the code into 6 different functions.

The core issue is that `get_job` has a lot of arguments. In order to test all possible combinations, this test takes 9 fixtures. The main test function actually has fewer than 2 lines per fixture, and I think has gotten to be about as efficient/readable as it can be while still fully testing the 9-argument function.

I squeezed out a little more readability by combining the secret-related checks into a helper.

The "baseline" for testing this was previously there were zero tests and people didn't seem to think having this test was necessary. Now there is a test that will find pretty much all the errors, though it's less than optimally readable (to be generous). It can probably be improved by splitting it somehow into multiple tests, but I think that's more effort than required here.